torch/csrc/lazy/core/lazy_graph_executor.cpp

```diff
@@ -468,8 +468,20 @@ void LazyGraphExecutor::SyncTensorsGraph(
   }
 }
 
-void LazyGraphExecutor::MarkStep(const BackendDevice& device) {
+struct MarkStepContext {
+  static bool in_context;
+  MarkStepContext() { in_context = true; }
+  ~MarkStepContext() { in_context = false; }
+};
+
+bool MarkStepContext::in_context = false;
+
+bool LazyGraphExecutor::InMarkStep(){ return MarkStepContext::in_context; }
+
+void LazyGraphExecutor::MarkStep(const BackendDevice& device, const std::vector<std::string>& devices, bool wait) {
   TORCH_LAZY_COUNTER("MarkStep", 1);
+  auto context = MarkStepContext();
```

Comment:
I am mixed about this approach because `Mark_step` means different things in python and C++. In Python it means execute the pending graph, truncate the IR and move on.

In C++, `Mark_step` just means we increment the step counter and do some cleanup. `SyncTensor`(execute the graph and truncate the IR graph) and `GetTensor`(execute the graph but don't truncate the IR graph) is the right level of the abstraction for the graph compilation/execution. I am a bit uncomfortable to push the high level concept(through a context object) of `python calls a mark_step` into the lower level implementation code.

One thing that might be slightly better is to pass `force_ltc_data` to `Compiles`
https://github.com/pytorch/pytorch/blob/52af4fc5baee363bb9c3170017d46d4fe0069c56/torch/csrc/lazy/core/lazy_graph_executor.cpp#L907

and you can use this to infer if callsite is from `mark_step`. @wconstab any thought?

Reply:
> I am mixed about this approach because Mark_step means different things in python and C++

Perhaps this is something we should unify on. I feel like it only causes confusion having the meaning in python and C++ be different. This is something that I've kind of touched up on by moving the sync tensor to inside `LazyGraphExecutor::MarkStep`. Unless there is some greater reason why the two different meanings were introduced that I'm missing?

> One thing that might be slightly better is to pass `force_ltc_data` to `Compiles`

Its not *just* the compiles though. It would have to be passed into the lowering context and then further the lowering context needs to pass it into each `Computation` object which is what the backend interface's `Compile` receives. There's just many more levels of indirection with this approach.

Having said that, if this is still preferred, I can make that change and that should still work.

Reply:
I think the right thing to do is to add a 
```
if(!config.force_ltc_data) {
  backend->HandleUnsyncCompilation()
}
```
but that would create one more backend interface which might be annoying.

Reply:
hmm, I like this idea. Its less intrusive and we can add a default implementation so that its not as annoying

Reply:
or, perhaps instead of a new method, we can add a parameter to the backend interface's compile?
```
getBackend()->Compile({computation}, config.force_ltc_data)
```
Maybe a bit more intrusive, but cleaner? Thoughts?

Reply:
In a normal case ie
```
t1 = .....
xm.mark_step()
print(t1)
```

print will trigger a `getTensor`-> `syncTensor(force_ltc_data=False)` and return in 
https://github.com/pytorch/pytorch/blob/ce9a477fdfb75adce8a3edb4b2feab698700e24d/torch/csrc/lazy/core/lazy_graph_executor.cpp#L893-L895
because every tensor is already materialized, no computation(compilation/execution) is needed.

in a case of 
```
t1 = .....
print(t1)
```
it will trigger a compilation and execution.

The up side of doing `getBackend()->Compile({computation}, config.force_ltc_data)` is we don't need to add a new backend API. the downside is it does not catch the case where the compilation is ccahed. it will return in 
https://github.com/pytorch/pytorch/blob/ce9a477fdfb75adce8a3edb4b2feab698700e24d/torch/csrc/lazy/core/lazy_graph_executor.cpp#L902-L905

This case is less annoying but still incur unintended execution but it is much faster than `compilation+execution`. Depending on which case you care about and how you want to handle them. For PyTorch/XLA I think we will just add a warning message so either approach works for me. I would also like to know what's Will's opinion regarding this.

Reply:
If there is no compilation, then I don't know that there's an issue. At least on our side. The extra compilation is the reason I brought this up as an issue. If it happens that the compile cache is hit, then it should not matter.

Reply:
The latest implementation looks good to me. @wconstab ?

Reply:
I think the objective is good, but i'm not sure if I understand the design choices.

Why is the warning issued from inside the backend layer instead of from inside lazy tensor core?